### ISSUE TITLE
Return of window borders

### ISSUE THREAD

**Issue by koekeishiya** [koekeishiya/yabai#565]
So I have recently due to a discussion in a different OSS application been able to come across some information that I believe will be able to solve *all* the issues with window borders that I talked about in https://github.com/koekeishiya/yabai/issues/487

My basic testing so far has made me quite optimistic that this is now achievable.

This will require the scripting-addition to be installed, but this is once more a feature I think is worth that trade-off. Apologize in advance to people who cannot run with SIP disabled for whatever reason, you can still use *[limelight](https://github.com/koekeishiya/limelight)* instead (a much lower quality version).

**Comment by tuananh** [comment]
This is awesome! Thanks

**Comment by itgoyo** [comment]
awesome work! Thanks

**Comment by burlapsax** [comment]
This is terrific @koekeishiya, but I am wondering: is there a way to prevent the borders being drawn when a window goes into native fullscreen mode? that feels pretty distracting IMHO, especially in cases like maximizing a video.

**Comment by glepnir** [comment]
Oh, welcome to back border!

**Comment by koekeishiya** [comment]
Borders are no longer drawn in native-fullscreen spaces (with the exception of Safari videos, because they apparently work differently). Also improved efficiency of resizing operations.

**Comment by malfario** [comment]
Awesome, thank you so much!

**Comment by burlapsax** [comment]
> Borders are no longer drawn in native-fullscreen spaces (with the exception of Safari videos, because they apparently work differently). Also improved efficiency of resizing operations.
> 
> Edit: Don't forget to reinstall the scripting addition (again), for these changes.

@koekeishiya truly outstanding! thank you 🙌

**Comment by rross101** [comment]
Borders looking great here.

**Comment by cmacrae** [comment]
This is great news!

For anyone using the Nix package/module who wants to test this out, you can use an overlay with `overrideAttrs` to follow master. Just be sure to `sudo yabai --install-sa` manually after for this to work:
```nix
{
  yabai = super.yabai.overrideAttrs (o: {
    version = "master";
    src = super.fetchFromGitHub {
      owner = "koekeishiya";
      repo = "yabai";
      rev = "17fcfb73d8bc013b0250a9be42adb91c1a7cb72e";
      sha256 = "065qdf5q955jr2cic47w0nxmp8n13dvjpmi6b779kggr38b1l7wz";
    };
  });
}
```

So far so good, the performance is much better! Thanks very much for your endeavours @koekeishiya 💜

**Comment by asierralozano** [comment]
Hello! 
I've updated yabai to the latest version following the wiki. I'm trying to test this new border feature, but it is not working for me. I've set the following lines within `.yabairc`

```
yabai -m config window_border on
yabai -m config window_border_width 10
yabai -m config active_window_border_color 0xff775759
yabai -m config normal_window_border_color 0xff00FFFF
```

I've also updated the scripting addition with `sudo yabai --install-sa` and restarted.
Any idea of what could be happening?

Thanks!

**Comment by koekeishiya** [comment]
Borders can now be toggled per window and specified to be enabled/disabled through rules regardless of the window_border setting. This should make it easier for people who may only want borders for some specific windows.

**Comment by yanzhang0219** [comment]
Hi @koekeishiya,
Is it possible to toggle border for a space? I have a keybinding to toggle a space between bsp and float, and what I want is that when I toggle a space to float, the borders of all the windows in this space will be off. Thank you very much.

**Comment by dominiklohmann** [comment]
> Is it possible to toggle border for a space like `yabai -m space --toggle border`? Thank you very much.

```bash
yabai -m query --windows --space |
  jq '.[].id' |
  xargs -I{} yabai -m window {} --toggle border
```

**Comment by yanzhang0219** [comment]
Hi @dominiklohmann, Thank you. 

What my requirement is that when I toggle a space to float, all windows in this space will not have border. "All windows" here I mean: 
1. all existing windows in this space
2. all newly-created windows in this space
3. all windows sent to this space in the future. 

"1" could be achieved by using your script. For "2", "3", what I am thinking is to write a script. When I create a new window or send a window to a space, I first check whether the type of the space is float or not (via `yabai -m query --spaces --space | jq .type`). If it is float, I will toggle its border via `yabai -m window --toggle border`. Am I correct?

However, this way looks burdensome. I'm wondering whether yabai could add the value `border` to `--toggle` of space like `yabai -m space --toggle border` so that after setting this, "1", "2", and "3" above will be implemented. @koekeishiya 

Thank you.

**Comment by dominiklohmann** [comment]
You surely could implement that using the signals API, but it's far from a trivial task.

Have you considered turning window borders on/off _globally_ depending on what space you're in, i.e., toggling them globally based on the signal `event=space_changed`?

**Comment by yanzhang0219** [comment]
Thank you. But you get me wrong (I updated my comment above). Briefly, my scenario is like this:
I am working in space no. 3 and suddenly I want to change this space to perform as plain macOS, i.e., floating and no border. And later, I may change it back. To implement this, I just bind a key to `yabai -m space --layout float && yabai -m space --toggle border`, and that's it.

**Comment by yanzhang0219** [comment]
My implementation so far is:

Step1. In skhdrc
```bash
alt + ctrl - f : yabai -m space --layout float && \
                 yabai -m config window_border off
alt + ctrl - b : yabai -m space --layout bsp && \
                 yabai -m config window_border on
```
When I change the space layout to float, I turn off the global border. And when I change the space layout back to bsp, I turn on the global border.

Step2. In yabairc
``` bash
yabai -m signal --add event=space_changed action="type=\$(yabai -m query --spaces --space | jq .type) && [ \$type = '\"bsp\"' ] && yabai -m config window_border on"
yabai -m signal --add event=space_changed action="type=\$(yabai -m query --spaces --space | jq .type) && [ \$type = '\"float\"' ] && yabai -m config window_border off"
```
When switch space, if the target space is bsp, turn on the global border, and if the target space is float, turn off the global border. 

My implementation above basically meets my requirement, but it has a drawback. That is when I switch a float space, I can see the border momently before it is off.

**Comment by tuananh** [comment]
border doesn't seem to work consistently for me.

for new windows, it's ok. click somewhere else and the border becomes like this

![image](https://user-images.githubusercontent.com/627278/84967443-23818180-b13e-11ea-8083-9af44a1ccb2b.png)

**Comment by jzelner** [comment]
Having the exact same issue.

**Comment by tuananh** [comment]
> Having the exact same issue.

you need to disable SIP and install sa. i updated my comment above.

**Comment by jzelner** [comment]
Oops - my apologies. I had done that but then not restarted the Dock...

**Comment by Aleppi** [comment]
Is it only for me that the borders don't disappear in native fullscreen, and that the borders cover parts of the app? For example the left most letter in Alacritty is partially underneath the border.

**Comment by cmacrae** [comment]
@Aleppi I have observed the border being drawn on native fullscreen applications, but only after starting/restarting yabai when the window is already in that state.

Regarding Alacritty, I think that's just the nature of where Alacritty is drawing its characters. You can address this by setting `window.padding.x` (there's also `window.padding.y`). I have both mine set to `15`, and my windows look like this:
![image](https://user-images.githubusercontent.com/3392199/85114961-d3282380-b212-11ea-85c6-7e6fcb224757.png)

**Comment by Aleppi** [comment]
@cmacrae For me it draws a border over fullscreen apps regardless if they were fullscreen before or after restarting yabai, but on my laptop it doesn't happen. It is not only alacritty where the overlap happens, it happens in all apps, for example the leftmost part of Discord where one sees if there are any notifications for a server is partially covered by the bar.

**Comment by sluukkonen** [comment]
Hey, I'd just like to give a big thank you for re-introducing window borders. My productivity took a big hit when they disappeared and I am delighted to see them return. Thank you so much for your hard work!

**Comment by Liquidmantis** [comment]
I've just started using Yabai today, and have been working on getting my config built out.  I've disabled SIP, installed the scripting additions, killed Dock, but still have an issue with the borders getting stuck.  If I maximize a window to native-fullscreen a seemingly random normal (not active) border from another window is stuck.  The only way I've been able to make it go away is to toggle borders.
